text/0000-prelude-2021.md

```diff
@@ -0,0 +1,172 @@
+[reference-level-explanation]: #reference-level-explanation
+
+The new prelude will be named after the edition that introduces it to make the link to the edition more obvious. Each edition will have a separate prelude, whether the contents are actually different or not.
+
+Importantly, we'll need to make the upgrade path to an edition containing new traits in the prelude as smooth as possible. To that end, cargo fix will update code that calls `try_from()` or `try_into()` in a scope that does not have the std traits in scope to use the explicit disambiguation syntax.
+
+Currently, diagnostics for trait methods that are not in scope suggest importing the originating trait. For traits that have become part of the prelude in a newer edition, the diagnostics should be updated such that they suggest upgrading to the latest edition in addition to importing the relevant trait.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Making the prelude contents edition-dependent makes the difference between different editions larger, and could thus increase confusion especially in the early phase when older editions and newer ones are used in parallel.
+
+Adding more traits to the prelude makes methods from other traits using the same names as prelude traits harder to access, requiring calls to explicitly disambiguate (`<usize as try_from::TryFrom>::try_from()`).
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## `TryFrom`/`TryInto`
+
+The `TryFrom`/`TryInto` traits could not be added to the prelude without an edition boundary. This was [tried](https://github.com/rust-lang/rust/pull/49305) around the time of the 2018 edition, but failed due to breaking substantial parts of the ecosystem. However, `TryFrom`/`TryInto` offer a generic conversion mechanism that is more robust than `as` because the conversions are explicitly fallible. Their usage is more widely applicable than the `From`/`Into` traits because they account for fallibility of the conversions.
+
+Without doing this, the `TryFrom`/`TryInto` traits remain less accessible than the infallible `From`/`Into` conversion traits, providing a disincentive to implement/use fallible conversions even where conversion operations are intrinsically fallible.
+
+## `FromIterator`
+
+The documentation for `FromIterator`'s `from_iter()` method currently reads:
+
+> `FromIterator::from_iter()` is rarely called explicitly, and is instead used through `Iterator::collect()` method. See `Iterator::collect()`'s documentation for more examples.
+
+However, it is reasonably common that type inferencing fails to infer the full type of the target type, in which case an explicit type annotation or turbofishing is needed (such as `iter.collect::<HashMap<_, _>>()` -- the type of the iterator item is available, so wildcards can be used for this). In these cases, `HashMap::from_iter(iter)` can be a more concise and readable way to spell this, which would be easier if the trait was in scope.
```

Comment:
I my experience `collect` is very often (guesstimate 90%) used at the end of a chain of method calls that may involve iterator adapters such as `map` etc. So replacing `collect` with `from_iter` would cause the entire chain to be nested in another set of parenthesis and likely indented to another level by rustfmt, which I find reads less nicely.  Even with `FromIterator` in the prelude I think I’d usually stick with `collect`, either with a turbofish or with a type annotation on the `let` statement when there’s one (guesstimate 70% of the time, sometimes just because splitting out a long-ish method call chain into a separate statement reads better).

```rust
let foo: Vec<_> = bar
    .iter()
    .flatten()
    .chain(baz)
    .cloned()
    .collect();
```

Reply:
Yeah, I don't believe that occasionally saving two characters and bringing another problem is big enough advantage to justify importing whole trait in prelude. I have same experience with it being used almost exclusively with combinators.